Feature: Do not default the player name to "Player"

rubidium42 opened the pull request:
## Motivation / Problem

Reviewing #9067 I noticed a "validation" of the client name getting added twice (after figuring out the UI breaks when no player name is set), looking at it the same validation also happens in other places. However, the rules for a valid client name differ between the different implementations, and there are many paths where the validation is not done.

So effectively this should solve the problem of players joining with either no name (which becomes "Player" when joining a server, but a self-hosted server will then have a client without name) or a "invalid" name. A name is invalid when it is empty of starts with spaces (that breaks tab completion).
In the places where the user interacts with OpenTTD the spaces will automatically be trimmed, however when receiving a client name over the network that is not done. After all, the name should be valid by that point so it is a 'broken' client.


## Description

Add a function (of authority) to validate a client name and use that everywhere where the client name is checked.
Add a (helper) function to check whether a client name is right, and if not show an error message. Every location where you would try to "join" a server (includes starting your own server) and where you change your name this check is performed.
The server will now reject people joining with invalid names and will kick people changing their name to an invalid name. With this rejection it changes the network protocol, but this validation is performed after the version check so it is inconsequential (unless clients/servers fudge their version).
Add trimming of the spaces at the begin and end of the client name; includes a few new functions in string.cpp.


## Limitations

I would be nicer for the user to get an UI where they can fix their user name in. However, that is quite complicated to accomplish as you would need to create a new window for that with the appropriate infrastructure to continue at the locations where you bailed out. The requires some significant rework of the code for relatively little gain, as for the end user the warning is usually shown in the Multiplayer window where you already have the edit box to enter it. The error message even tells the user where that edit box is.
Other cases where this could be triggered are when joining from the command line. Most likely you already set your nick before, so it is unlikely to happen there. Finally is changing the setting. You do that on the console and immediately get an error message (also on the console) that the name was wrong, so you can fix it there.

It does not solve the current "Player" issue, as that name is still allowed. However, for new installs that name will not be prefilled anymore so it should slowly die out.

The edit box in the Multiplayer UI does not perform direct validation as that is quite a nuisance and clearing the filled in value would then already show an error message, making it quite a hassle. Especially as the chance that they do not fill anything in that edit box when they change it is likely quite small, whereas players joining MP for the first time are far more likely to just click join without thinking about the player name.


## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, gs_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

rubidium42 committed:
Codechange: [Network] Introduce function to validate the client name

rubidium42 committed:
Feature: [Network] Ensure players fill in a name instead of defaulting to "Player"

rubidium42 committed:
Change: [Network] Prevent invalid client names being sent to the server when changing it using the console/settings

rubidium42 committed:
Add: [Network] Validate the client name server side, so no clients with invalid names can actually join

TrueBrain commented on src/console_cmds.cpp:
```diff
@@ -716,6 +716,11 @@ DEF_CONSOLE_CMD(ConClientNickChange)
 	}
 
+	if (!NetworkIsValidClientName(argv[2])) {
+		IConsoleError("Cannot give a client an invalid name");
```
Can we make this error a bit nicer? At least give some snippets what "invald" is? It now sounds so harsh :D

rubidium42 replied:
I'm a bit conflicted about this. For the end user it might be nicer, though development wise any changes made in the NetworkIsValidClientName require this message to be changed as it might otherwise be out-of-date. Having said that, entering a (currently) invalid name already requires the server owner knowing that you need to use two double quotes to pass an empty name, and double quotes with spaces between to get a name starting with a space.
To make matters "worse", I'd argue most changes of other people's nicks are done via rcon which requires the command to be put in double quotes, so you cannot even use double quotes for the client name with rcon making it impossible to trigger it with the current implementation of de validation.
So the only case where this error could be triggered is when you are hosting a game yourself (non-dedicated) and want to change the nick of some other client, which to me sounds like the demographic that does not know about the console.

LordAro commented:
I'd fully expect the game to trim leading/trailing spaces for me, no need to error for that usecase

rubidium42 committed:
Add: String functionality to trim spaces from C-style strings

rubidium42 commented:
> I'd fully expect the game to trim leading/trailing spaces for me, no need to error for that usecase

Just implemented that, with a new function for trimming strings.

TrueBrain commented on src/network/network_client.cpp:
```diff
@@ -1258,6 +1295,9 @@ void NetworkUpdateClientName()
 
 	if (ci == nullptr) return;
+	/* There is no validation on string settings. So, when the client name is changed via the console
+	 * there is no easy way to prevent an invalid name. Though, we can prevent it getting sent here. */
```
Is this true? You can have validations and post-functions on settings that change. Can't we do these things there?

TrueBrain replied:
https://github.com/OpenTTD/OpenTTD/blob/31c5b8fe0f9c070143625478f7120e0c9c2332fd/src/settings.cpp#L1327 for example, I think.

rubidium42 replied:
This function is called in the post-function of the setting.
Yes, I can do this validation there. However, then I would be putting more networking logic in the post-function that is going to call this function, so putting it here feels better to me. Now it is also only going to complain when it is actually used.

TrueBrain commented on src/network/network_server.cpp:
```diff
@@ -1441,6 +1444,11 @@ NetworkRecvStatus ServerNetworkGameSocketHandler::Receive_CLIENT_SET_NAME(Packet
 
 	if (ci != nullptr) {
+		if (!NetworkIsValidClientName(client_name)) {
+			/* An invalid client name was given. This was already validated by the client. */
```
"This was already validated by the client" <- what do you mean?

rubidium42 replied:
I've changed the comment to be less terse:
An invalid client name was given. However, the client ensures the name is valid before it is sent over the network, so something went horribly wrong. This is probably someone trying to troll us.

TrueBrain commented on src/string.cpp:
```diff
@@ -302,6 +302,62 @@ bool StrValid(const char *str, const char *last)
+		*str = '\0';
+	} else {
+		last_non_space[1] = '\0';
+	}
```
This if feels weird, but I cannot fully put my finger on it. `last_non_space` smaller than `str` means that there are trailing whitespaces not? So shouldn't this be inverted?
Also, I only expect a single if, as in, no else. As if there is no trailing whitespace, nothing changes, right?

TrueBrain commented on src/network/network_client.cpp:
```diff
@@ -1250,6 +1251,42 @@ void NetworkClientsToSpectators(CompanyID cid)
+ */
+bool NetworkShowErrorMessageForInvalidClientName(char *client_name)
+{
+	if (!StrEmpty(client_name)) StrTrimInPlace(client_name);
```
I have been looking at this for a good 10 minutes now. I get why this line is here, but it also feels weird.

Basically, this function only does validation, but sneaky also trims the string if needed. This means the client name is always validated properly and fixed if possible, but the timing seems a bit weird. At least it is not something I would have expected this function to do (it is also not documented it does :P).

So it made me think, can't we do this different. For the popup in my Network GUI, we can. As there we can show the error as soon as you changed the name that it is invalid, or trim it. For the console, we also trim it. Via settings we can do via a change-callback. That only leaves 2 places where it is more difficult:

- Startup with an invalid name.
- The main network GUI where the player name is just an edit box.

Startup should be easy to solve; pretty sure we have more code doing that. Maybe even the change-callback is called on load? I don't even know :D

The main network GUI is a bit more tricky. I would suggest either of the two solutions:
- Do like the new Network GUI, and only show the name, with an edit icon next to it to edit it. This gives a Query window, making it better to show the error after pressing OK
- Do the trimming after editing, but leave the error for later. That at least means this trimming can be removed from here, and put close to the places that manipulate it.

What do you think?
(sorry for the long story, but I didn't just want to say: I think this is wrong :D)

rubidium42 replied:
> Startup should be easy to solve; pretty sure we have more code doing that. Maybe even the change-callback is called on load? I don't even know :D

I think we should not require that someone enters their name just to start the game. Only once you start using the network should it ask for a name.

> The main network GUI is a bit more tricky. I would suggest either of the two solutions:
> * Do like the new Network GUI, and only show the name, with an edit icon next to it to edit it. This gives a Query window, making it better to show the error after pressing OK
> * Do the trimming after editing, but leave the error for later. That at least means this trimming can be removed from here, and put close to the places that manipulate it.

The "fun" with the edit box is that it calls OnEditboxChanged after every key stroke. Now if you want to write your full name (given and surname) with a space in between, every time you type the space just before entering your last name it will be trimmed. That's going to cause a lot of bug reports, so sadly the trimming can only be done once we are fairly certain the user is done editing, which coincidentally is the moment they try to join a game or start a server.

Hiding the input field for the user name might actually make it harder for new users to find the place where to enter their name.

TrueBrain reviewed (commented):
I think auto-trimming is an awesome move :D That is really good!

rubidium42 committed:
Add: [Network] Validate the client name when receiving one from the server

This so names from other clients are known valid in the client as well, instead allowing some compromised/bad server to potentially crash clients upon certain expectations.

TrueBrain reviewed (approved):
Assuming you find no more bugs yourself

LordAro commented:
There's more than one sort of whitespace, and though it'd be very difficult to get a tab, newline, carriage return or vertical tab character into that edit window, I'd expect a basic "trim" function to filter out all whitespace characters, not just spaces

rubidium42 commented:
> There's more than one sort of whitespace, and though it'd be very difficult to get a tab, newline, carriage return or vertical tab character into that edit window, I'd expect a basic "trim" function to filter out all whitespace characters, not just spaces

That could indeed be an expectation, but filtering out all whitespace becomes complex really fast. First, you need to decode the UTF-8 to a wide char, then you need to check for a white space character using iswspace. That function is locale dependent so it behaves differently for different people. The only thing that is certain is that it returns true when isspace returns true.

Then there's also that the string validation routines filter out all the special characters you mentioned. Since those characters cannot be entered, why add logic for detecting them while trimming?

So practically this would make the trimming significantly more complex with the UTF8 decode and locale dependent determination whether something is a space. Or I would be replacing the `*c == ' '` with `isspace(*c)`, where anything but `' '` would already be removed by the string validation of OpenTTD.
